src/Control/FunFlow/Base.hs

```diff
@@ -14,20 +16,48 @@ import qualified Control.FunFlow.ContentStore    as CS
 import           Control.FunFlow.Diagram
 import           Control.FunFlow.External
 import qualified Control.FunFlow.External.Docker as Docker
+import           Data.ByteString                 (ByteString)
+import           Data.Default
 import           Data.Proxy                      (Proxy (..))
 import           Data.Store
 import qualified Data.Text                       as T
 import           Prelude                         hiding (id, (.))
 
+-- | A cacher is responsible for controlling how steps are cached.
+data Cacher i o =
+    NoCache -- ^ This step cannot be cached (default).
+  | Cache
+    { -- | A unique identifier included in the cache key to ensure
+      --   that the same input does not give the same cache key for
+      --   different steps.
+      cacherUniqueIdent :: Int
+    , cacherKey         :: i -> Int
```

Comment:
Why not re-use `ContentHash` for this?

Reply:
Hmm. I can't easily use the typeclass without making it a constraint on the `Flow'` constructor, which screws up `NoCache` (we could do stuff with `Dict`, but that seems overcomplicated). I guess we could use `i -> ContentHash`, which would be fine, but currently we can only construct them in `IO`.

Maybe let's turn the question around: what is the advantage of using `ContentHash` here? This seems to be a case closer to traditional `Hashable` to me.

Reply:
AFAIK the `Hashable` hash does not have strong guarantees on avoiding hash collisions.
For hashsets/tables that's fine, they use an additional `Eq` constraint to handle collisions.
What's the mechanism here?

With `ContentHash` were already relying on collisions being sufficiently unlikely.
The underlying implementation of `ContentHashable` could be pure in most cases. The `IO` is just for file access. I was wondering before if it would make sense to have something like

```
class Monad m => ContentHashable m a where
    contentHashUpdate :: Context -> a -> m ContentHash
```

`Hashable` has the advantage of being less heavy weight, and having more instance builtin.
However, I think we'd have to handle collisions in that case.